lib/matplotlib/category.py

```diff
@@ -1,123 +1,162 @@
-        vmap = dict(zip(axis.unit_data.seq, axis.unit_data.locs))
+        # dtype = object preserves numerical pass throughs
+        values = np.atleast_1d(np.array(value, dtype=object))
 
-        if isinstance(value, six.string_types):
-            return vmap[value]
+        # pass through sequence of non binary numbers
+        if all((units.ConversionInterface.is_numlike(v) and
+                not isinstance(v, VALID_TYPES)) for v in values):
+            return np.asarray(values, dtype=float)
 
-        vals = shim_array(value)
+        # force an update so it also does type checking
+        unit.update(values)
 
-        for lab, loc in vmap.items():
-            vals[vals == lab] = loc
+        str2idx = np.vectorize(unit._mapping.__getitem__,
+                               otypes=[float])
 
-        return vals.astype('float')
+        mapped_value = str2idx(values)
+        return mapped_value
 
     @staticmethod
     def axisinfo(unit, axis):
-        majloc = StrCategoryLocator(axis.unit_data.locs)
-        majfmt = StrCategoryFormatter(axis.unit_data.seq)
+        """Sets the axis ticks and labels
+        """
+        # locator and formatter take mapping dict because
+        # args need to be pass by reference for updates
+        majloc = StrCategoryLocator(axis.units)
+        majfmt = StrCategoryFormatter(axis.units)
         return units.AxisInfo(majloc=majloc, majfmt=majfmt)
 
     @staticmethod
     def default_units(data, axis):
-        # the conversion call stack is:
+        # the conversion call stack is supposed to be
         # default_units->axis_info->convert
-        if axis.unit_data is None:
-            axis.unit_data = UnitData(data)
+        if axis.units is None:
+            axis.set_units(UnitData(data))
         else:
-            axis.unit_data.update(data)
-        return None
+            axis.units.update(data)
+        return axis.units
 
 
-class StrCategoryLocator(ticker.FixedLocator):
-    def __init__(self, locs):
-        self.locs = locs
-        self.nbins = None
+class StrCategoryLocator(ticker.Locator):
+    """tick at every integer mapping of the string data"""
+    def __init__(self, units):
+        """
+        Parameters
+        -----------
+        units: dict
+              (string, integer) mapping
+        """
+        self._units = units
 
+    def __call__(self):
+        return list(self._units._mapping.values())
 
-class StrCategoryFormatter(ticker.FixedFormatter):
-    def __init__(self, seq):
-        self.seq = seq
-        self.offset_string = ''
+    def tick_values(self, vmin, vmax):
+        return self()
 
 
-class UnitData(object):
-    # debatable makes sense to special code missing values
-    spdict = {'nan': -1.0, 'inf': -2.0, '-inf': -3.0}
+class StrCategoryFormatter(ticker.Formatter):
+    """String representation of the data at every tick"""
+    def __init__(self, units):
+        """
+        Parameters
+        ----------
+        units: dict
```

Comment:
This isn't right either, is it?  Its a UnitData object as well, right?

Reply:
yeah, updating this now